#54 Docker update

microwavekonijn opened the pull request:
This update will give us some benefits. Now the Dockerfile is the same for testing and running the bot. Also it makes smart use of the cache when building containers, making rebuild times of containers a lot faster. Also it will auto start the bot when the container starts.

Issue #54

microwavekonijn committed:
Refactor of Docker

- Bot will automatically run on start container
- Ignore node_modules and local.json when building
- Tests use the regular Dockerfile, given more consistency
- Travis installs the missing dev dependencies
- Bot starts now by calling npm start
- Production and staging compose use now the build image from Docker Hub

microwavekonijn committed:
Copy the local.json config on local environments

microwavekonijn committed:
Fix and added comments

- Building the container will only install normal dependencies
- Added comments for clarification

microwavekonijn committed:
Fixes

- npm-debug.log ingnore by npm and docker
- Added scripts to npm for local testing and running using a docker container
- Fixed container naming in docker-compose for production

JamesLongman commented on package.json:
```diff
@@ -4,7 +4,10 @@
   "main": "index.js",
   "scripts": {
-    "test": "mocha"
+    "start": "node src/bot.js",
```
Let's swap ```start``` with ```start:docker``` and ```test``` with ```test:docker``` to make this mildly simpler

microwavekonijn replied:
I want to leave it likes this. test is used by travis directly and start is used by the container, I just added start:docker and test:docker to make running those container easier for our own benefit.

JamesLongman replied:
Can we not just swap the values in references? I'm just thinking the local version is going to be the one we're typing and I'd rather type start 5000 times than start:docker

microwavekonijn replied:
You can use up arrow to reuse previous commands you entered. Also in ide's(at least Webstorm) you can add a script which you can run with the press of a button. Honestly I would like to look at a solution(and I know it is possible) where it detects changes and then restarts the bot with the new code.

Besides, instead of building a container each time we make changes(even with use of the docker cache now) we should really look into just running the bot locally without docker. The only problem is that one of the dependencies uses python and needs to compile some stuff (and I have no idea how to fix it). To make sure that we run the bot with the intended node version we can use nvm, tho running it with a newer version should not bring any problems (just it will support newer features, which will fail during testing when implemented).

**Edit:** Added up and down scripts.

JamesLongman replied:
As the section currently stands if we change ```test``` --> ```mocha``` and ```test:docker``` --> ```test``` I think this would be good to merge.

Automatic rebuilding would be cool, but running the bot outside of our docker environment would be very bad for consistency. The whole point of docker setup is to hopefully provide identical environments on all systems. You are of course free to do whatever you want locally but I would recommend against it

microwavekonijn replied:
Differences between platforms and versions are very minimal as far as I can tell. We should probably set some requirement conditions for which the bot should work with (like the nodejs version). We can always expand it when we find problems, but we don't really want to trade ease of development for avoidance of possible problems we might encounter. Also testing should catch any problems and else be expanded to catch those problems. I already encountered this when I used finally on a promise.

JamesLongman commented on package.json:
```diff
@@ -4,7 +4,10 @@
-    "test": "mocha"
+    "start": "node src/bot.js",
+    "test": "mocha",
+    "start:docker": "docker-compose up --build -d",
```
```docker-compose up --build``` without the detached mode would be better as this is what we'll be using locally in most instances

microwavekonijn replied:
Just one question, we only use one container now. How will this work when we have multiple containers(like digbot + database)? Will it console log everything from all containers?

microwavekonijn replied:
I renamed some commands and added a few:
- docker:start will build and start the containers in detached mode
- docker:logs will give a realtime view of the digbot containers logs
- docker:stop will stop the containers and remove them from docker

The reason behind this is that when you run it without detach and use ctl+c to stop the container it will exit with an error and you still have to use the down command. An extra command could be added to combine start and logs scripts.

JamesLongman replied:
Well before I believe it was just run from a separate directory, so detached or not it wouldn't have mattered, I don't remember though.

My main complaint on the function of the scripts is the lack of logs so if we could just have an ```up``` script to bring it up with logs and ```down``` to bring it down that could work

microwavekonijn replied:
Added those script.

JamesLongman commented on docker-compose-staging.yml:
```diff
@@ -2,15 +2,12 @@
         environment:
         - NODE_ENV=staging
-        - DIGBOT_PORT=1338
+        - DIGBOT_PORT=1337
```
What's up with this? You changed this in the config refactor to 1338 and now back to 1337?

microwavekonijn replied:
Yeah, I copied the old config and there it was port 1338 in staging and then for the env vars I just copied it there without much thought. Then I saw that the port which is forwarded is 1337 so I made them match again.

Thinking about it we can probably define a default port in default.json and then change the port in the yml file in the port section. Then the only purpose for the DIGBOT_PORT var is when you want to change the port and not running it in a docker container.

The only question I have, what is the port for? I think it has something to do with jenkins. If it is used don't you think it should be forwarded in the Dockerfile by default?

JamesLongman replied:
I have no idea what this port is/was for. If I had to guess I would say Jenkins, or maybe it logs into the API and listens on this port, no clue without digging into it

JamesLongman commented on .dockerignore:
```diff
@@ -0,0 +1,3 @@
+node_modules
```
You musn't be afraid to dream a little bigger darling, my proposed changes:

```
# Copyright © 2018 DIG Development team. All rights reserved.

# Ignore local npm files (node_modules will still exist in docker as it will create its own)
node_modules
npm-debug.log

# Ignore local.json files to exclude from images
config/local.json

# Ignore all md files
*.md

# Ignore .git files, .eslintrc.json, .travis.yml, etc
.*

# Ignore docker file and docker compose files
Dockerfile
docker-compose*
```

.*, in particular, will yield a substantial benefit as git files can really grow; the only thing that concerns me is the dokerfile docker-compose section as I'm not sure how this will affect how we're going to be deploying with docker hub. Worth giving it a try though

JamesLongman commented on docker-compose.yml:
```diff
@@ -2,11 +2,11 @@
+    digbot:
+        build: .
+        volumes:
+        # Links the local.json config to the container for local development
```
embedding here

microwavekonijn replied:
??

JamesLongman replied:
Embedding of some options in inconsistent across our docker compose files, specifically, docker-compose-test.yml has 4 space embedding for lines 11 and 12 for volumes: values but this file does not

JamesLongman reviewed (changes requested):
Definately more efficient build order and better use of compose files. The way npm install is split is very efficient.

Needs docs to explain ```npm run-script x```

Our current docs need to be improving not disapearing, docker/README.md can be moved or its information can be put in the main readme but it can't just be deleted.

Usually I would say we need to make sure we're including project file changes in tests but #60 is blocking that so no need

microwavekonijn committed:
Amended the dockerignore and package files

- .dockerignore is mad at some files
- renamed some script and added some more

microwavekonijn committed:
Added documentation and refactor

- Added some documentation with requirements and instructions
- Added default env vars to the Dockerfile
- The port in the container is now exposed by default to port 1337, this can be overwritten by mapping the port to another port on the host

microwavekonijn committed:
Fixed small error

microwavekonijn committed:
I am a horrible writer

microwavekonijn committed:
Added 36 spaces

microwavekonijn committed:
Added up and down scripts

- npm run up is alias for npm run docker:start && npm run docker:logs
- npm run down is alias for npm run docker:stop

JamesLongman commented on README.md:
```diff
@@ -99,6 +99,22 @@ DIGBot is capable of reacting when events are scheduled to remind people of thei
 Some channels the community would prefer to always have a free version of. This is the case for channels such as dedicated voice channels for certain games, for example if a member joins a voice channel for a game such as planetside DIGBot will react by creating a second voice channel for that game so there is still a free channel. Every-time a modular channel become populated or empty, DIGBot wi ...
 
 
+## Usage
```
This I'm happy to alter on develop

microwavekonijn replied:
It definitely needs some improvement. I added it to at least have something on how to run it as the docker readme files was removed.

microwavekonijn committed:
Small fixes and addition to readme

microwavekonijn committed:
Renamed npm scripts

microwavekonijn committed:
Fix CI

JamesLongman commented on package.json:
```diff
@@ -4,7 +4,14 @@
-    "test": "mocha"
+    "start": "node src/bot.js",
+    "mocha": "mocha",
+    "test": "docker-compose -f docker-compose-test.yml up --build -d && docker-compose exec digbot npm install --only=dev && docker-compose exec digbot npm test && docker-compose down",
```
calling itself, update npm test to npm mocha?

microwavekonijn replied:
Whoops :o

microwavekonijn committed:
Completely forgot this one
